### ISSUE TITLE
sampleAnnotation fails due to various reason after update to 1.0.0

### ISSUE THREAD

**Issue by nicweb** [gagneurlab/drop#114]
Hi,

I'm encountering various problems trying to run the DROP pipeline.

When there are there is no external count matrix defined the sampleAnnotation step fails with

"single positional indexer is out-of-bounds"

I'm also encountering problems as the DROP_GROUP is recognized as boolean and is therefore missing str values and so on...

Steps to reproduce:
stage demo
Delete 2 last lines from sample_annotation.tsv from the demo package.
run snakemake sampleannotation

**Comment by mumichae** [comment]
Hi,

Thanks for reporting this bug.
I could reproduce it and have implemented a fix in Pull request #116. Once it has been merged, you can update drop in your environment via the dev branch as described in the [documentation](https://gagneurlab-drop.readthedocs.io/en/latest/installation.html#other-drop-versions)

**Comment by mumichae** [comment]
This issue has now been resolved, and it's available via the `dev` branch. Please let me know if it fixes your issue.

**Comment by nicweb** [comment]
I can not confirm this is fixed.
I updated to the latest dev version - removing the latest last two lines of the sample_annotation.tsv it now fails with a key error for the group configured in the aberrantExpression section.

**Comment by nicweb** [comment]
In addition in my own config I also can not run the pipeline as the groups are still parsed as boolean instead of strings..

**Comment by mumichae** [comment]
Could  you give me an example of the group names in your  config and sample annotation? What error message do you get?

**Comment by mumichae** [comment]
Also, when you update drop,  please also run
```
drop update
```
In your project directory. This is a bit inconvenient atm, will look into automatically updating drop if the version changes in the next release.

**Comment by nicweb** [comment]
I run "drop update" after upgrading to the latest dev version. Can you confirm that the the clean demo runs through without the last 2 lines? For me this can easily reproduced with the latest code: 



> snakemake sampleAnnotation --cores 1
> check for missing R packages
> KeyError in line 13 of /tests/demo4/Snakefile:
> 'import_exp'
>   File "//tests/demo4/Snakefile", line 13, in <module>
>   File "/home/centos/miniconda3/lib/python3.7/site-packages/drop/config/DropConfig.py", line 50, in __init__
>   File "/home/centos/miniconda3/lib/python3.7/site-packages/drop/config/submodules/AberrantExpression.py", line 21, in __init__
>   File "/home/centos/miniconda3/lib/python3.7/site-packages/drop/config/submodules/AberrantExpression.py", line 21, in <dictcomp>`


My groups are 

> cut -f 5 sample_annotation.tsv
> DROP_GROUP
> outrider,mae,fraser,blood
> outrider,mae,fraser
> outrider,mae,fraser
> outrider,mae,fraser
> outrider,mae,fraser
> outrider,mae,fraser
> outrider,mae,fraser
> outrider,mae,fraser
> outrider,mae,fraser
> outrider,mae,fraser
> outrider,mae,fraser
> outrider,mae,fraser
> outrider,mae,fraser
> outrider,mae,fraser
> outrider,mae,fraser
> outrider,mae,fraser
> outrider,mae,fraser,blood
> outrider,mae,fraser
> outrider,mae,fraser
> outrider,mae,fraser
> outrider,mae,fraser
> outrider,mae,fraser
> outrider,mae,fraser
> outrider,mae,fraser
> outrider,mae,fraser,blood
> outrider,mae,fraser
> outrider,mae,fraser
> outrider,mae,fraser
> outrider,mae,fraser
> outrider,mae,fraser
> outrider,mae,fraser
> outrider,mae,fraser
> outrider,mae,fraser
> outrider,mae,fraser`

and I'll get the error:



> snakemake sampleAnnotation --cores 1
> check for missing R packages
> AttributeError in line 13 of /tests/rna2/Snakefile:
> Can only use .str accessor with string values!
>   File "/tests/rna2/Snakefile", line 13, in <module>
>   File "/home/centos/miniconda3/lib/python3.7/site-packages/drop/config/DropConfig.py", line 43, in __init__
>   File "/home/centos/miniconda3/lib/python3.7/site-packages/drop/config/SampleAnnotation.py", line 24, in __init__
>   File "/home/centos/miniconda3/lib/python3.7/site-packages/drop/config/SampleAnnotation.py", line 45, in parse
>   File "/home/centos/miniconda3/lib/python3.7/site-packages/pandas/core/generic.py", line 5135, in __getattr__
>   File "/home/centos/miniconda3/lib/python3.7/site-packages/pandas/core/accessor.py", line 187, in __get__
>   File "/home/centos/miniconda3/lib/python3.7/site-packages/pandas/core/strings.py", line 2100, in __init__
>   File "/home/centos/miniconda3/lib/python3.7/site-packages/pandas/core/strings.py", line 2157, in _validate

Edit: Version I use:

> pip freeze | grep drop
> drop @ git+https://github.com/gagneurlab/drop.git@4fcd140cf549b6e768a4f79547a7130b04998a8e

**Comment by mumichae** [comment]
Oh I must have missed actually running the code after removing the lines. I addressed this in #122.
```
pip install git+https://github.com/mumichae/drop.git@fix_import_counts
```
If you just remove the last 2 lines in the sample annotation, you should get an error saying that you don't have enough samples per group. To avoid this, simply remove `import_exp` from the `groups` of `aberrantExpression` in the config file.

**Comment by mumichae** [comment]
As for the `DROP_GROUP` parsing, I'm afraid I can't reproduce the error with data you provided. #122 now also includes column type checking of the sample annotation file upon reading it. Maybe that'll fix the issue.

Otherwise, you can check if the following works
```
import pandas as pd
sa = pd.read_csv("sample_annotation.tsv", sep="\t")
print(sa["DROP_GROUP"])
print(col.str.replace(" ", "").str.replace("(|)", "", regex=True))
```

**Comment by gevro** [comment]
I think my issue #125 is the same problem.

**Comment by mumichae** [comment]
@gevro yes,  please try to update drop as above to see if it fixes the issue

**Comment by gevro** [comment]
Tried updating now, but problem persists.

**Comment by nicweb** [comment]
The update fixes the initial index out of bound error. Now I'm still stuck at the Key error.
Additionally your code snippet reports `Name: DROP_GROUP, dtype: bool`
the second print fails as col is not defined...

**Comment by mumichae** [comment]
Could it be that the DROP_GROUP column is empty? Did you also get the error with the demo project? Check the complete sa object whether it represents what you have in your sample annotation file.

I forgot to add
```
col = sa["DROP_GROUP"]
```
But the last command would fail of course

**Comment by gevro** [comment]
On my system, DROP_GROUP is filled for every sample. I tried your code snipped and did not get any errors:
>>> print(sa["DROP_GROUP"])
0          GTEX100,GTEXALL
1          GTEX100,GTEXALL
2          GTEX100,GTEXALL
3          GTEX100,GTEXALL
4          GTEX100,GTEXALL
              ...         
676    GTEX100,GTEXALL,UDP
677    GTEX100,GTEXALL,UDP
678    GTEX100,GTEXALL,UDP
679    GTEX100,GTEXALL,UDP
680    GTEX100,GTEXALL,UDP
Name: DROP_GROUP, Length: 681, dtype: object
>>> print(col.str.replace(" ", "").str.replace("(|)", "", regex=True))
0          GTEX100,GTEXALL
1          GTEX100,GTEXALL
2          GTEX100,GTEXALL
3          GTEX100,GTEXALL
4          GTEX100,GTEXALL
              ...         
676    GTEX100,GTEXALL,UDP
677    GTEX100,GTEXALL,UDP
678    GTEX100,GTEXALL,UDP
679    GTEX100,GTEXALL,UDP
680    GTEX100,GTEXALL,UDP
Name: DROP_GROUP, Length: 681, dtype: object

**Comment by nicweb** [comment]
I managed to produce a minimal example:

Attached test.txt is a minimal sample_annotation.tsv. with 1 line containing demo files resulting in the key error in a newly staged demo folder.
Hopefully you are able to reproduce the error.

[test.txt](https://github.com/gagneurlab/drop/files/5475612/test.txt)

Maybe we should split the issue into the drop group error and the key error to make the discussion easier?

**Comment by nicweb** [comment]
The drop group is not empty:

>  print(sa["DROP_GROUP"])
> I18R008Rf02    True
> ...
> I18R008Ri03    True
> Name: DROP_GROUP, dtype: bool
> ...
> print(col.str.replace(" ", "").str.replace("(|)", "", regex=True))
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
>   File "/home/centos/miniconda3/envs/drop/lib/python3.8/site-packages/pandas/core/generic.py", line 5135, in __getattr__
>     return object.__getattribute__(self, name)
>   File "/home/centos/miniconda3/envs/drop/lib/python3.8/site-packages/pandas/core/accessor.py", line 187, in __get__
>     accessor_obj = self._accessor(obj)
>   File "/home/centos/miniconda3/envs/drop/lib/python3.8/site-packages/pandas/core/strings.py", line 2100, in __init__
>     self._inferred_dtype = self._validate(data)
>   File "/home/centos/miniconda3/envs/drop/lib/python3.8/site-packages/pandas/core/strings.py", line 2157, in _validate
>     raise AttributeError("Can only use .str accessor with string values!")
> AttributeError: Can only use .str accessor with string values!

Last line failure is expected as this is no string. Maybe this is a panda version error?

**Comment by mumichae** [comment]
Agreed, please refer to #126

**Comment by mumichae** [comment]
Missing key error when no import group is assigned is now fixed in #122 cd22254a0e2f36de9a400c7efaa02694e1105d2c
